src/burst_buffer/dws.lua

```diff
@@ -0,0 +1,227 @@
+	yaml = string.gsub(yaml, "WLMID", wlmID)
+	yaml = string.gsub(yaml, "JOBID", jobID)
+	yaml = string.gsub(yaml, "USERID", userID)
+	yaml = string.gsub(yaml, "GROUPID", groupID)
+	self.yaml = yaml
+end
+
+-- DWS.save will save the text of the Workflow resource into a file
+-- of the given name.
+-- On success this returns true.
+-- On failure this returns false and an optional error message.
+function DWS:save(fname)
+	local f = io.open(fname, "w")
+	if f == nil 
+	then
+		local msg = "unable to open " .. fname
+		return false, msg
+	end
+	f:write(self.yaml)
+	local rc = {f:close()}
+	-- Success or failure is a boolean in rc[1].
+	return rc[1]
+end
+
+-- DWS.apply will apply the Workflow resource via kubectl.
+-- On success this returns true.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:apply(fname)
+	return self:io_popen("kubectl apply -f " .. fname .. " 2>&1")
+end
+
+-- DWS.delete will delete the named Workflow resource via kubectl.
+-- On success this returns true.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:delete()
+	return self:io_popen("kubectl delete workflow " .. self.name .. " 2>&1")
+end
+
+-- DWS.get_jsonpath will get the specified values from the Workflow resource.
+-- On success this returns true and a string containing the result of the query.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:get_jsonpath(path)
+	local cmd = "kubectl get workflow " .. self.name .. " -o jsonpath='" .. path .. "' 2>&1"
+	return self:io_popen(cmd)
+end
+
+-- DWS.get_current_state will get the status of the Workflow resource with
+-- respect to its desired state.
+-- On success this returns true and an array containing the desiredState,
+-- the current state, and the status of the current state.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:get_current_state()
+	local ret, output = self:get_jsonpath([[{.spec.desiredState}{"\n"}{.status.state}{"\n"}{.status.status}{"\n"}]])
+	if ret == false then
+		return ret, output
+	end
+	local status = {}
+	local idx = 0
+	for line in output:gmatch("[^\n]+") do
+		status[idx] = line
+		idx = idx + 1
+	end
+	if idx == 1 and status[0] == "proposal" then
+		-- DWS has not yet attached a status section to this new
+		-- resource.  Fill out the table for our consumer.
+		status[1] = ""
+		status[2] = ""
+	elseif idx ~= 3 then
+		-- We should have had 3 lines.
+		ret = false
+	end
+	return ret, status
+end
+
+-- DWS.get_hurry will get the hurry flag of the Workflow resource.
+-- On success this returns true and a boolean for the value of the hurry flag.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:get_hurry()
+	local ret, result = self:get_jsonpath("{.spec.hurry}")
+	if ret ~= true then
+		return ret, result
+	end
+	local hurry = false
+	if result == "true" then
+		hurry = true
+	end
+	return true, hurry
```

Comment:
```suggestion
        if result == "true" then
                return true, true
        else
                return ret, result
        end
```

Reply:
I think this function should be rewritten to return `true` in the first position if the hurry flag is set, and then an error string (or nil) in the second position if there's an issue.

Reply:
In all of these, the first return value is a bool indicating whether the kubectl command succeeded.  I don't want this one to deviate from that pattern.

Reply:
In that context, ignore my suggestion. It's better as written.